src/core/renderer_vk/renderer_vk.cpp

```diff
@@ -0,0 +1,546 @@
+
+	instanceInfo.ppEnabledExtensionNames = instanceExtensions.data();
+	instanceInfo.enabledExtensionCount = instanceExtensions.size();
+
+	if (auto createResult = vk::createInstanceUnique(instanceInfo); createResult.result == vk::Result::eSuccess) {
+		instance = std::move(createResult.value);
+	} else {
+		Helpers::panic("Error creating Vulkan instance: %s\n", vk::to_string(createResult.result).c_str());
+	}
+	// Initialize instance-specific function pointers
+	VULKAN_HPP_DEFAULT_DISPATCHER.init(instance.get());
+
+	// Enable debug messenger if the instance was able to be created with debug_utils
+	if (std::find(
+			instanceExtensions.begin(), instanceExtensions.end(),
+			// std::string_view has a way to compare itself to `const char*`
+			// so by casting it, we get the actual string comparisons
+			// and not pointer-comparisons
+			std::string_view(VK_EXT_DEBUG_UTILS_EXTENSION_NAME)
+		) != instanceExtensions.end()) {
+		vk::DebugUtilsMessengerCreateInfoEXT debugCreateInfo{};
+		debugCreateInfo.messageSeverity = vk::DebugUtilsMessageSeverityFlagBitsEXT::eVerbose | vk::DebugUtilsMessageSeverityFlagBitsEXT::eInfo |
+										  vk::DebugUtilsMessageSeverityFlagBitsEXT::eError | vk::DebugUtilsMessageSeverityFlagBitsEXT::eWarning;
+		debugCreateInfo.messageType = vk::DebugUtilsMessageTypeFlagBitsEXT::ePerformance | vk::DebugUtilsMessageTypeFlagBitsEXT::eValidation |
+									  vk::DebugUtilsMessageTypeFlagBitsEXT::eGeneral;
+		debugCreateInfo.pfnUserCallback = &Vulkan::debugMessageCallback;
+		if (auto createResult = instance->createDebugUtilsMessengerEXTUnique(debugCreateInfo); createResult.result == vk::Result::eSuccess) {
+			debugMessenger = std::move(createResult.value);
+		} else {
+			Helpers::warn("Error registering debug messenger: %s", vk::to_string(createResult.result).c_str());
+		}
+	}
+
+	// Create surface
+	if (VkSurfaceKHR newSurface; SDL_Vulkan_CreateSurface(window, instance.get(), &newSurface)) {
+		surface.reset(newSurface);
+	} else {
+		Helpers::warn("Error creating Vulkan surface");
+	}
+
+	// Pick physical device
+	if (auto enumerateResult = instance->enumeratePhysicalDevices(); enumerateResult.result == vk::Result::eSuccess) {
+		std::vector<vk::PhysicalDevice> physicalDevices = std::move(enumerateResult.value);
+		std::vector<vk::PhysicalDevice>::iterator partitionEnd = physicalDevices.end();
+
+		// Prefer GPUs that can access the surface
+		const auto surfaceSupport = [this](const vk::PhysicalDevice& physicalDevice) -> bool {
+			const usize queueCount = physicalDevice.getQueueFamilyProperties().size();
+			for (usize queueIndex = 0; queueIndex < queueCount; ++queueIndex) {
+				if (auto supportResult = physicalDevice.getSurfaceSupportKHR(queueIndex, surface.get());
+					supportResult.result == vk::Result::eSuccess) {
+					return supportResult.value;
+				}
+			}
+			return false;
+		};
+
+		partitionEnd = std::stable_partition(physicalDevices.begin(), partitionEnd, surfaceSupport);
+
+		// Prefer Discrete GPUs
```

Comment:
Unlike OpenGL, Vk lets us control device creation very well, so maybe we could add a confic about whether to prefer discrete vs integrated in the future (laptop users who don't want their laptop to run warm or eat up battery would like it maybe)

Reply:
The next step I had in mind with all this "stable_partition" stuff is to have a bit of config-syntax to define the device-preference. So it will order the device-list partitioning/sorting based on a passed-in list of criteria like `devicePriority = "discrete>apiVersion>vram>"`, or whatever that config-syntax would be, to avoid having to have a whole potentially unstable device UUID defined in a config file.
The more share-able and stand-alone config files are the better.